module/zfs/zvol.c

```diff
@@ -735,34 +742,42 @@ zvol_discard(struct bio *bio)
 static int
 zvol_read(struct bio *bio)
 {
+	uio_t uio;
 	zvol_state_t *zv = bio->bi_bdev->bd_disk->private_data;
-	uint64_t offset = BIO_BI_SECTOR(bio) << 9;
-	uint64_t size = BIO_BI_SIZE(bio);
-	int error;
+	uint64_t volsize = zv->zv_volsize;
 	rl_t *rl;
-	uio_t uio;
+	int error = 0;
 
-	if (size == 0)
+	uio.uio_resid = BIO_BI_SIZE(bio);
+
+	if (uio.uio_resid == 0)
 		return (0);
 
 	uio.uio_bvec = &bio->bi_io_vec[BIO_BI_IDX(bio)];
 	uio.uio_skip = BIO_BI_SKIP(bio);
-	uio.uio_resid = size;
 	uio.uio_iovcnt = bio->bi_vcnt - BIO_BI_IDX(bio);
-	uio.uio_loffset = offset;
+	uio.uio_loffset = BIO_BI_SECTOR(bio) << 9;
 	uio.uio_limit = MAXOFFSET_T;
 	uio.uio_segflg = UIO_BVEC;
 
-	rl = zfs_range_lock(&zv->zv_znode, offset, size, RL_READER);
+	rl = zfs_range_lock(&zv->zv_znode, uio.uio_loffset, uio.uio_resid,
+	    RL_READER);
+	while (uio.uio_resid > 0 && uio.uio_loffset < volsize) {
+		uint64_t bytes = MIN(uio.uio_resid, DMU_MAX_ACCESS >> 1);
 
-	error = dmu_read_uio(zv->zv_objset, ZVOL_OBJ, &uio, size);
+		/* don't read past the end */
+		if (bytes > volsize - uio.uio_loffset)
+			bytes = volsize - uio.uio_loffset;
 
+		error = dmu_read_uio_dbuf(zv->zv_dbuf, ZVOL_OBJ, &uio, bytes);
```

Comment:
Don't need ZVOL_OBJ as per function:
dmu_read_uio_dbuf(dmu_buf_t *zdb, uio_t *uio, uint64_t size)

error = dmu_read_uio_dbuf(zv->zv_dbuf, &uio, bytes);
That should be correct.